### PR TITLE
Remove lookup-tx-by-utxo functionality

### PR DESCRIPTION
**Pull request by sipa**
There are 3 ways through which main.cpp (and by extension, the `getrawtransaction` and REST tx lookup) find transactions:
- In the mempool (works always)
- In the block index (works only when -txindex is enabled).
- In the UTXO set (inefficient, and only for confirmed transactions that have some of their outputs not yet spent by other confirmed transactions)

This third option is confusing because the conditions under which it works are complicated, and can make it seem like it always works even without -txindex. Remove it. If you need to call `getrawtransaction` for confirmed transactions, you almost certainly need -txindex.

This only impacts the `getrawtransaction` and REST tx calls, so document those.

### CODE REVIEW

**Comment by laanwj** [comment]
What is the migration path for users that were using this to query the utxo set?
I suppose `gettxout` and REST `/rest/getutxos` cover that?

**Comment by sipa** [comment]
I guess I'm making an (unsubstatiated) claim/guess that this functionality
is so hard to rely on that there are no users.

I'll gladly reconsider if I'm wrong.

**Comment by sipa** [comment]
Yes, use UTXO query functions to query the UTXO set.

**Comment by laanwj** [comment]
This definitely needs mention in the release notes.
Maybe deprecate this functionality in 0.13 (put it behind an option) then actually remove it in 0.14?

**Comment by laanwj** [comment]
```
abandonconflict.py:
Initializing test directory /tmp/testvmbco_b4/4
start_node: bitcoind started, waiting for RPC to come up
start_node: RPC succesfully started
start_node: bitcoind started, waiting for RPC to come up
start_node: RPC succesfully started
JSONRPC error: No information available about transaction. Use `gettransaction` for wallet transactions, or enable -txindex to use getrawtransaction on confirmed transactions.
Stopping nodes
Not cleaning up dir /tmp/testvmbco_b4/4
Failed
stderr:
   File "/home/travis/build/bitcoin/bitcoin/bitcoin-i686-pc-linux-gnu/qa/rpc-tests/test_framework/test_framework.py", line 144, in main
    self.run_test()
  File "/home/travis/build/bitcoin/bitcoin/bitcoin-i686-pc-linux-gnu/qa/rpc-tests/abandonconflict.py", line 42, in run_test
    nA = next(i for i, vout in enumerate(self.nodes[0].getrawtransaction(txA, 1)["vout"]) if vout["value"] == Decimal("10"))
  File "/home/travis/build/bitcoin/bitcoin/bitcoin-i686-pc-linux-gnu/qa/rpc-tests/test_framework/coverage.py", line 49, in __call__
    return_val = self.auth_service_proxy_instance.__call__(*args, **kwargs)
  File "/home/travis/build/bitcoin/bitcoin/bitcoin-i686-pc-linux-gnu/qa/rpc-tests/test_framework/authproxy.py", line 144, in __call__
    raise JSONRPCException(response['error'])
Pass: False, Duration: 6 s
```

**Comment by dcousens** [comment]
concept ACK

**Comment by gmaxwell** [comment]
I agree that it's so incomprehensible (and not desirable) in its behavior that it's unlikely anyone was using it (or at least, using it _and_ not silently getting screwed over by it).  I do wish any of these queries that only worked  with txindex had a non-txindex mechanism that took a location parameter (e.g. block height.)

**Comment by MarcoFalke** [comment]
I don't think we need to deprecate this behind a command line flag. (Keep in mind that there still would be no way for us to find out how many would toggle the flag)

Considering that this is not really a functionality someone can rely on and it appears unlikely someone (other than our test framework) relies on it, it is probably safe to just remove it.

**Comment by laanwj** [comment]
> I don't think we need to deprecate this behind a command line flag. (Keep in mind that there still would be no way for us to find out how many would toggle the flag)

That's true, but it would give people the opportunity to complain about it, while not having to forego the upgrade.

I don't feel strongly about it though, if you guys believe that this is a change that can be made without warning in advance I'm fine with that. Depends on how urgent this change is as well.

> I do wish any of these queries that only worked with txindex had a non-txindex mechanism that took a location parameter (e.g. block height.)

One radical idea would be to expose CDiskTxPos fields on the interface here and there, and add a REST/RPC call to get a transaction from such a (blockfile_id,offset) tuple. I wonder if that would make some things easier/possible, such as external indices without having to store the block chain twice, but with fast access. Probably these 'pointers' should be obfuscated somehow to discourage direct access to the block files from outside bitcoind.

**Comment by sipa** [comment]
> Depends on how urgent this change is as well.

The changes here are not urgent. If we choose to not adopt them quickly, I think at least the updated error message in `getrawtransaction` is useful.

**Comment by petertodd** [comment]
One of my clients has something that was using this FWIW, so I'd target actual removal for v0.14

**Comment by laanwj** [comment]
> One of my clients has something that was using this FWIW, so I'd target actual removal for v0.14

That's interesting. I somehow expected that, if some functionality is available for so long, some people will come to rely on it.
What is their use case? Could they (in principle, given warning upfront to update their code) use the other UTXO query commands as replacement?

**Comment by petertodd** [comment]
@laanwj Basically their wallet code happens to depend on it; with a fair amount of new code they could work around the problem (remember that simply saving the txs when created isn't enough due to malleability).

I don't think that's a show-stopper, but we should give users some kind of warning period. Perhaps we could make v0.13.0 *appear* to not have this support by default, but then re-enable it with a command line flag, that's marked as "will be removed in v0.14.0"?

**Comment by laanwj** [comment]
>  Perhaps we could make v0.13.0 appear to not have this support by default, but then re-enable it with a command line flag, that's marked as "will be removed in v0.14.0"?

+1. That's exactly what I suggested too:

> Maybe deprecate this functionality in 0.13 (put it behind an option that is disabled by default) then actually remove it in 0.14?

**Comment by petertodd** [comment]
@laanwj Cool, sounds like we're in agreement. :)

**Comment by sipa** [comment]
Ok I'll reword this to just be the extra RPC help/error, and a message that
it is deprecated?

**Comment by petertodd** [comment]
@sipa Sounds good to me, thanks!

**Comment by laanwj** [comment]
Looks like this causes a few RPC tests to fail:
```
wallet.py                      | False  | 9 s
rest.py                        | False  | 7 s
abandonconflict.py             | False  | 7 s
```

**Comment by dcousens** [comment]
utACK c3af143,  but needs the command line (or maybe just a compile flag?)

**Comment by laanwj** [comment]
Still failing travis.

**Comment by gmaxwell** [comment]
@sipa Needs rebase.

**Comment by morcos** [comment]
I'm ever so slightly against this change.  I find it's routinely useful for me when I'm trying to debug behavior on a random node.

**Comment by sipa** [comment]
@morcos Perhaps that's a good reason to introduce a `getmempooltransaction` RPC? My reason for this (or a related) change is that `getrawtransaction` has very unclear conditions under which it works, and I keep seeing people say it doesn't work.

**Comment by laanwj** [comment]
> Perhaps that's a good reason to introduce a getmempooltransaction RPC

Agree with @morcos that the way toward fixing this could also be make the API clearer instead of removing functionality.

+1 for splitting up `getrawtransaction` into calls that look in one place only.
